theseus/extlib/cusolver_lu_solver.cpp

```diff
@@ -0,0 +1,334 @@
+
+    CUSOLVER_CHECK(cusolverRfBatchAnalyze(cusolverRfH));
+
+    P = P_cpu.cuda();
+    Q = Q_cpu.cuda();
+}
+
+std::vector<int> CusolverLUSolver::factor(const torch::Tensor& A_val) {
+
+    TORCH_CHECK(A_val.device().is_cuda());
+    TORCH_CHECK(A_val.dim() == 2);
+    TORCH_CHECK(A_val.size(0) <= batchSize); // currently failing unless "==", cublas bug?!
+    TORCH_CHECK(A_val.size(1) == nnz);
+
+    factorId++;
+    factoredBatchSize = A_val.size(0);
+
+    at::Tensor A_val_array_cpu = torch::empty(factoredBatchSize * sizeof(double*), torch::TensorOptions(torch::kByte));
+    double* pA_val = A_val.data_ptr<double>();
+    double** pA_val_array_cpu = (double**)A_val_array_cpu.data_ptr<uint8_t>();
+    for(int i = 0; i < factoredBatchSize; i++) {
+        pA_val_array_cpu[i] = pA_val + nnz * i;
+    }
+    at::Tensor A_val_array = A_val_array_cpu.cuda();
+
+    CUSOLVER_CHECK(cusolverRfBatchResetValues(factoredBatchSize,
+                                              rows, nnz,
+                                              A_rowPtr.data_ptr<int>(), A_colInd.data_ptr<int>(), (double**)A_val_array.data_ptr<uint8_t>(),
+                                              P.data_ptr<int>(), Q.data_ptr<int>(),
+                                              cusolverRfH));
+
+    CUSOLVER_CHECK(cusolverRfBatchRefactor(cusolverRfH));
+
+    std::vector<int> singularityPositions(factoredBatchSize);
+    CUSOLVER_CHECK(cusolverRfBatchZeroPivot(cusolverRfH, singularityPositions.data()));
+    for(int i = 0; i < factoredBatchSize; i++) {
+        if (singularityPositions[i] >= 0){
+            fprintf(stderr, "Error: A[%d] is not invertible, singularity=%d\n", i, singularityPositions[i]);
+        }
+    }
+
+    return singularityPositions;
+}
+
+void CusolverLUSolver::solve(const torch::Tensor& b) {
+
+    TORCH_CHECK(b.device().is_cuda());
+    TORCH_CHECK(b.dim() == 2);
+    TORCH_CHECK(b.size(0) == factoredBatchSize);
+    TORCH_CHECK(b.size(1) == rows);
+
+    at::Tensor b_array_cpu = torch::empty(factoredBatchSize * sizeof(double*),
+                                          torch::TensorOptions(torch::kByte));
+    double* pB = b.data_ptr<double>();
+    double** pB_array_cpu = (double**)b_array_cpu.data_ptr<uint8_t>();
+    for(int i = 0; i < factoredBatchSize; i++) {
+        pB_array_cpu[i] = pB + rows * i;
+    }
+    at::Tensor b_array = b_array_cpu.cuda();
+    at::Tensor temp = torch::empty(rows * 2 * factoredBatchSize,
+                                   torch::TensorOptions(torch::kDouble).device(A_rowPtr.device()));
+
+    CUSOLVER_CHECK(cusolverRfBatchSolve(cusolverRfH,
+                                        P.data_ptr<int>(), Q.data_ptr<int>(),
+                                        1, // nrhs
+                                        temp.data_ptr<double>(), rows,
+                                        (double**)b_array.data_ptr<uint8_t>(), rows));
+}
+
+PYBIND11_MODULE(cusolver_lu_solver, m) {
```

Comment:
Over time we will likely add more things to extlib. Does it make sense to keep the pybind stuff in a separate file?